### PR TITLE
TY: fix TyProjection folding

### PR DESCRIPTION
**Pull request by vlad20012**
bors r+

### CODE REVIEW

**Comment by vlad20012** [comment]
fixed tests

File: src/main/kotlin/org/rust/lang/core/types/BoundElement.kt [hunk 1]
```diff
@@ -52,7 +52,7 @@ data class BoundElement<out E : RsElement>(
         )
 
     override fun superVisitWith(visitor: TypeVisitor): Boolean = assoc.values.any { visitor.visitTy(it) } ||
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/lang/core/types/BoundElement.kt]: Shouldn't we use `visitWith` instead of `visitTy` and `visitRegion` here?

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/types/BoundElement.kt]: Fixed

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 2]
```diff
@@ -52,13 +56,21 @@ sealed class RsDiagnostic(
         element: PsiElement,
         private val expectedTy: Ty,
         private val actualTy: Ty
-    ) : RsDiagnostic(element, inspectionClass = RsTypeCheckInspection::class.java) {
+    ) : RsDiagnostic(element, inspectionClass = RsTypeCheckInspection::class.java), TypeFoldable<TypeError> {
+        private val description: String? = if (expectedTy.hasTyInfer || actualTy.hasTyInfer) {
+            // if types contain infer types, they will be replaced with more specific (e.g. `{integer}` with `i32`)
+            // so we capture string representation of types right here
+            expectedFound(expectedTy, actualTy)
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: Can you write a test for this case?

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: Hmm, this should be covered by multiple tests, but we don't have any tests that check such error description (of any annotation), so if we want to start testing it, maybe we can do it in a separate PR?

**Pull request merged**